Real-Time Data RFC

ibgreen opened the pull request:
# For #3072 

#### Background
- Data polling support for layers

#### Change List
- New RFC

Pessimistress commented:
I don't think the existing implementation for the scene graph example is unreasonably complex. If I have a real-world app that requires polling, it's likely that

- I need to process the data in some fashion (e.g. apply it as a patch) rather than replacing the existing table entirely.
- I need the data for something more than just this layer, which means I'll load and store it externally.
- I will implement some logic that dynamically controls the next reload (instruction included in the server response, or based on the size of the diff, etc.)

This example is an extremely simple version of such use cases, and insufficient to inform the design of a new API.

ibgreen commented:
Thanks for the feedback,

> I don't think the existing implementation for the scene graph example is unreasonably complex.

Whether we should keep simplifying the API for various use cases is a fair question. I have hoped that we'd keep refining the API to work in a declarative way for more use cases, improving support in JSON API examples, for new users etc. 

I also fel that stating improved real-time support would also be a nice way to market our next framework release.

> If I have a real-world app that requires polling, it's likely that

> I need to process the data in some fashion (e.g. apply it as a patch) rather than replacing the existing table entirely.

Yes, this is already supported, I added back this example to the RFC, I cut it initially to simplify.

> I need the data for something more than just this layer, e.g. shared with other layers or charting components, which means I need to load and store the data externally.

I feel that is a separate concern from this RFC as it really applies to any use of the `data URL` prop. I'd be happy to start a separate "data sharing RFC" to discuss this. 

(A natural way to address that would be to e.g. define a separate set of top-level `dataSources` (like mapbox does) (each with an `id` and then `id` can be referenced by layers), and perhaps improve facilities for data sharing in composite layers).

> I will implement some logic that dynamically controls the next reload (instruction included in the server response, or based on the size of the diff, etc.)

You can still do programmatically that of course. I added an example to the RFC. This RFC is intended to be minimal and the proposed API does not try to make those aspects declarative.

> This example is an extremely simple version of such use cases, 

While the use case is fairly simple, I think is the "80% use" case. The ability to regularly poll a simple data source and automatically update the vis. This is especially true if we provide "object key" support for transitions as mentioned in the RFC.

> and insufficient to inform the design of a new API.

There not that much of a new API to inform: The proposal is a single prop `dataRefreshIntervalMs` that activates automatic reloads. 

Even if we wanted to make significant future refinements to handle other real-time use cases, how likely is it that we would feel that adding the `dataRefreshIntervalMs` prop was a mistake?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/deck.gl?pullRequest=3073) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/deck.gl?pullRequest=3073) before we can accept your contribution.<br/><hr/>**Ib Green** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/deck.gl?pullRequest=3073) it.</sub>
